keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,965 @@
+it is conventiont to just initialize the global logger once and then have all
+code pick it up from there.
+
+The API is:
+
+```go
+// SetFallbackLogger may only be called once per program run. Setting it
+// multiple times would not have the intended effect because the new logger
+// would not be used by code which already retrieved the fallback logger
+// earlier.
+//
+// Therefore this also should be called before any code starts to use logging.
+// Such code will get a valid logger to avoid crashes, but that logger will add
+// "INIT-LOG-FIXME" as prefix to all output to indicate that SetFallbackLogger
+// should have been called first.
+func SetFallbackLogger(log Logger) {
+    if fallbackLoggerWasSet {
+        panic("the global logger was already set and cannot be changed again")
+    }
+    fallbackLogger = log
+    fallbackLoggerWasSet = true
+}
+
+// FromContext retrieves a logger set by the caller or, if not set,
+// falls back to the program's fallback logger.
+func FromContext(ctx context.Context) Logger {
+    if logger, err := logr.FromContext(ctx); err == nil {
+        return logger
+    }
+
+    return fallbackLogger
+}
+
+// TODO can be used as a last resort by code that has no means of
+// receiving a logger from its caller. FromContext or an explicit logger
+// parameter should be used instead.
+//
+// This function may get deprecated at some point when enough code has been
+// converted to accepting a logger from the caller and direct access to the
+// fallback logger is not needed anymore.
+func TODO() Logger {
+    return fallbackLogger
+}
+
+// Background retrieves the fallback logger. It should not be called before
+// that logger was initialized by the program and not by code that should
+// better receive a logger via its parameters. TODO can be used as a temporary
+// solution for such code.
+func Background() Logger {
+    return fallbackLogger
+}
+```
+
+#### logging helper API
+
+The following helper code will get copied from klog:
+- `ObjectRef`
+- `KMetaData`
+- `KObj`
+- `KRef`
+- `KObjs`
+
+To ensure that a single import is enough in source files, that package will also contain:
+
+```go
+type Logger = logr.Logger
+var NewContext = logr.NewContext
+// plus more as needed...
+```
+
+<<[UNRESOLVED @pohly ]>>
+
+TODO: this might be an opportunity to enhance performance of KObj without
+causing an API break in kobj - see https://github.com/kubernetes/klog/pull/261.
+
+<<[/UNRESOLVED]>>
+
+#### Logging in tests
+
+Ginkgo tests do not need to be changed. Because each process only runs one test
+at a time, the global default can be set and local Logger instances can be
+passed around just as in a normal binary.
+
+Unit tests with `go test` require a bit more work. Each test case must
+initialize a `logr/testing` Logger for its own instance of `testing.T`.
+
+<<[UNRESOLVED @pohly ]>>
+
+What is a good default log level for unit tests? Should JSON output be enabled
```

Comment:
If we make 0 the default log threshold, then only `logger.Info` and `logger.Error` messages will be visible in a failed `go test` test case. All log messages with higher verbosity like `logger.(V1).Info` will not be included.

This is not going to be useful for debugging the test failure and therefore I think we should use a higher threshold by default. Otherwise everyone calling `go test` (= developers, CI jobs) will have to remember to bump up the default.

Looking at https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md, I think level 5 would be a good default:

> Trace level verbosity
>
> * Context to understand the steps leading up to errors and warnings
> * More information for troubleshooting reported issues

Anything lower and we are back at the "must rerun test to collect more information" state.

Reply:
Ah, I misunderstood the paragraph: this is about the log level threshold.

In E2E tests, we are using 4, which is honestly very noisy, because we see the logs for even successful tests. I understand that we won't be changing this log level, as that runs in a different binary. But it serves as a precedent. 

In the case of kube-scheduler, I think 3 gives you a good balance.

Reply:
The beauty of contextual logging in this case is that you don't get the output of successful tests at all. `go test` collects the output of each test case in memory and only dumps it to stdout if a test fails, unless you use `go test -v`. Therefore the concern about 4 or 5 being too noisy doesn't apply here.